copyoptions.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"math"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/semaphore"
+)
+
+const defaultConcurrencyLimit = 3
+
+type copyOptions struct {
+	limiter             *semaphore.Weighted
+	copyCallbackHandler func(ctx context.Context, desc ocispec.Descriptor) error
+}
+
+func NewCopyOptions() *copyOptions {
+	return &copyOptions{
+		limiter: semaphore.NewWeighted(defaultConcurrencyLimit),
+	}
+}
+
+func (co *copyOptions) WithConcurrencyLimit(concurrency int64) *copyOptions {
+	co.limiter = semaphore.NewWeighted(concurrency)
+	return co
+}
+
+func (co *copyOptions) WithCopyCallbackHandler(fn func(ctx context.Context, desc ocispec.Descriptor) error) *copyOptions {
+	co.copyCallbackHandler = fn
+	return co
+}
+
+type extendedCopyOptions struct {
+	*copyOptions
+	depth               uint32
+	upEdgeFilterHandler func(ctx context.Context, desc ocispec.Descriptor) (bool, error)
+	customUpEdgesFinder func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error)
```

Comment:
naming: this is more like a `upEdgeFinder`

Reply:
A default `upEdgesFinder` will be used if this is not provided. So this is more like a "custom" one.